### ISSUE TITLE
Remove PIDs and introduce types and objects for processes

### ISSUE THREAD

**Issue by yorickpeterse** [inko-lang/inko#231]
## Summary

PIDs will be removed from processes, and processes will be accessible using
dedicated types/objects instead of PIDs. This allows for a more pleasant API,
and removes the need for generating PIDs concurrently.

## Motivation

The use of PIDs is something Inko inherited from Erlang/OTP, mostly because I
was thinking to add support for remote processes in the early days of Inko.
Since then I have realised that support for remote processes/nodes is something
better suited for third-party libraries. Any network based RPC would likely need
to be secured using SSL/TLS, require some serialisation protocol (something like
[BERT](http://bert-rpc.org/)), and possibly more. Baking all of this in the VM
is a lot of work, and likely won't satisfy everybody's needs. For this reason,
I never bothered with it (and never will).

This means we're left with using PIDs, but for no particular good reason.
Meanwhile, generating PIDs is expensive (especially when recycling PIDs) and
leads to a somewhat weird API. For example, today you can send messages like so:

```inko
import std::process

let pid = process.spawn {
  process.receive # => 'hello'
}

process.send(pid: pid, message: 'hello')
```

Here `pid` is an `Integer`, because PIDs internally are just integers. The above
API is dynamically typed, because we can't attach additional information (such
as the message type) to an `Integer`. To work around that, the
`std::process.channel` method was introduced. This method allows you to write
the above as follows:

```inko
import std::process

let sender = process.channel!(String) lambda (receiver) {
  receiver.receive
}

sender.send('hello')
```

While this API is type-safe, it's needlessly verbose. It's also a different API,
requiring users to figure out themselves which one they need to use.

We could merge the two APIs together to make it more pleasant to use, but
internally we'd still be using PIDs so basically it's just syntax sugar.

Generating PIDs as mentioned is also expensive. Allocating a PID is a multi-step
process that involves incrementing atomic integers, hash map lookups, and a hash
map insert.

Instead of keeping all of the above, we'll make two changes:

1. PIDs are removed entirely. This allows us to remove the global process table,
   and any lookups made using it. This should speed up spawning processes quite
   a bit, as no synchronisation will be necessary.
2. In the VM we'll add a process prototype and use this for the various process
   operations. 
3. In the runtime we'll introduce a new API that makes use of this setup.

With this new API, spawning processes would be done as follows:

```inko
let proc = Process.new do (proc) {
  proc.receive # => 'hello'
}

proc.send('hello')
```

This API is type-safe by default, instead of requiring a different API.

## Implementation

In the VM we'll need to introduce an additional prototype object to use for
processes, and all VM instructions for process operations need to be adjusted to
expect and use such an object, instead of using an integer.

When a process is spawned, we allocate an object for it in both the parent and
the newly spawned process. This object is then passed as the first argument to
the block that the process will run. This removes the need for having to send
the block over from the parent.

In the runtime, we'll need to introduce a `Process!(T)` type that wraps the VM's
process prototype object. Since processes use lambdas we'd normally have to use
the `lambda` keyword, but this becomes a bit obnoxious to type instead of `do`.
To work around this, we need to extend the compiler so that it can infer
closures that explicitly use `do` as lambdas _if_ they don't capture any
variables. This means the following two lines would be identical:

```inko
let proc = Process.new do (proc) {}
let proc = Process.new lambda (proc) {}
```

Making receiving a message type-safe requires some extra work. For example, the
following does not compile using Inko as it currently is in `master`:

```inko
object Process!(T) {
  # ...
}

let proc = Process.new do (proc) {
  proc.receive.to_uppercase
}

proc.send('hello')
```

This won't compile because the type of `proc.receive` is determined to be `T`,
as the body of the closure is type-checked before `proc.send('hello')` is
checked. There are two ways we could work around this:

The first option is that we define a type parameter on `Process.new` and require
developers to explicitly pass one:

```inko
object Process!(T) {
  def new!(V)(block: lambda (Process!(V))) -> Process!(V) {
    # ...
  }
}

let proc = Process.new!(String) do (proc) {
  proc.receive.to_uppercase
}

proc.send('hello')
```

This can be done today, but it's a bit verbose. We could also use
`Process!(String).new` and remove the need for defining a custom `new` method,
but this requires parser changes as `A!(B)` is currently a syntax error.

An alternative is that we extend the compiler so it can infer type parameters
based on the type they are casted to explicitly. This would let you write the
following:

```inko
object Process!(T) {
  def new!(V)(block: lambda (Process!(V))) -> Process!(V) {
    # ...
  }
}

let proc = Process.new do (proc) {
  let message = proc.receive as String

  message.to_uppercase
}

proc.send('hello')
```

This however requires compiler changes, and it might break type-checking in
other places if we're not careful to not overwrite concrete types already
associated with a type parameter. This setup could also result to types being
inferred differently if there are multiple `receive` calls and a developer
changes the order of these calls.

## Drawbacks

We'll most likely need to extend the compiler to make type inference smarter,
removing the need for manual type annotations as much as possible. Processes
will now also always heap allocate, and `std::process.current` would heap
allocate on every call (though we might later optimise that somehow).

For the main process there's also the problem that we don't know what the return
type of `std::process.current` should be, and we can't ensure it's always the
same type in the same process. Take this code for example:

```inko
import std::process

process.current.send('foo')
process.current.send(10)
```

If the return type of `process.current` is `Process!(T)`, both calls would be
considered valid. This is also true for newly spawned processes:

```inko
import std::process

Process.new do (proc) {
  process.current # => How would we know the return type?
}
```

In short: the return type of `std::process.current` would vary based on what
process it is executed in. We could return a `Dynamic` and require explicit
casting, but this would be no better than returning a `Process!(T)`.

## Related work

* [Pony](https://tutorial.ponylang.io/types/actors.html)
* [Alpaca](https://github.com/alpaca-lang/alpaca)
